File: python/oncell/runtime.py
"""Runtime — runs inside the cell. Loads the agent, handles requests.

The Host Agent starts this process inside the gVisor sandbox.
It listens on a unix socket for requests from the Host Agent,
dispatches them to the developer's Agent class, and streams responses back.

Usage (called by the base image entrypoint):
    python -m oncell.runtime --agent /app/agent.py --cell-id acme-corp --port 8080

Protocol:
    POST /setup           → calls agent.setup(ctx)
    POST /request         → calls agent.on_request(ctx, method, params)
    POST /teardown        → calls agent.teardown(ctx)
    GET  /health          → {"status": "ok"}
"""

from __future__ import annotations

import argparse
import asyncio
import importlib.util
import inspect
import json
import os
import signal
import sys
import traceback
from pathlib import Path
from typing import Any

from oncell.agent import Agent
from oncell.cell import Cell


def load_agent(agent_path: str) -> Agent:
    """Load an Agent subclass from a Python file.

    Finds the first class that subclasses Agent in the given file.
    """
    path = Path(agent_path).resolve()
    if not path.exists():
        raise FileNotFoundError(f"agent file not found: {agent_path}")

    spec = importlib.util.spec_from_file_location("_agent_module", str(path))
    if spec is None or spec.loader is None:
        raise ImportError(f"cannot load module from {agent_path}")

    module = importlib.util.module_from_spec(spec)
    sys.modules["_agent_module"] = module
    spec.loader.exec_module(module)

    # Find the Agent subclass
    agent_cls = None
    for name, obj in inspect.getmembers(module, inspect.isclass):
        if issubclass(obj, Agent) and obj is not Agent:
            agent_cls = obj
            break

    if agent_cls is None:
        raise ValueError(f"no Agent subclass found in {agent_path}")

    return agent_cls()


async def handle_request(agent: Agent, cell: Cell, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
    """Handle a single HTTP-like request over the connection."""
    try:
        # Read request line
        request_line = await reader.readline()
        if not request_line:
            return

        request_line = request_line.decode().strip()
        method, path, _ = request_line.split(" ", 2)

        # Read headers
        content_length = 0
        while True:
            line = await reader.readline()
            if line == b"\r\n" or line == b"\n" or not line:
                break
            header = line.decode().strip()
            if header.lower().startswith("content-length:"):
                content_length = int(header.split(":", 1)[1].strip())

        # Read body
        body = None
        if content_length > 0:
            body = await reader.readexactly(content_length)
            body = json.loads(body)

        # Dispatch
        status = 200
        result: Any = None

        if path == "/health":
            result = {"status": "ok", "cell_id": cell.id}

        elif path == "/setup":
            await agent.setup(cell)
            result = {"status": "ok"}

        elif path == "/request":
            if not body or "method" not in body:
                status = 400
                result = {"error": "missing 'method' in request body"}
            else:
                req_method = body["method"]
                params = body.get("params", {})
                result = await agent.on_request(cell, req_method, params)

        elif path == "/teardown":
            await agent.teardown(cell)
            result = {"status": "ok"}

        else:
            status = 404
            result = {"error": f"not found: {path}"}

    except Exception as e:
        status = 500
        result = {"error": str(e), "traceback": traceback.format_exc()}

    # Write response
    response_body = json.dumps(result, default=str).encode()
    response = (
        f"HTTP/1.1 {status} OK\r\n"
        f"Content-Type: application/json\r\n"
        f"Content-Length: {len(response_body)}\r\n"
        f"\r\n"
    ).encode() + response_body

    writer.write(response)
    await writer.drain()
    writer.close()
    await writer.wait_closed()


async def run_server(agent: Agent, cell: Cell, port: int):
    """Start the runtime HTTP server."""
    server = await asyncio.start_server(
        lambda r, w: handle_request(agent, cell, r, w),
        host="0.0.0.0",
        port=port,
    )

    print(f"oncell runtime started — cell={cell.id} port={port}", flush=True)

    # Start heartbeat
    cell.heartbeat.start()

    # Handle graceful shutdown
    stop = asyncio.Event()

    def on_signal():
        print("shutting down...", flush=True)
        stop.set()

    loop = asyncio.get_event_loop()
    for sig in (signal.SIGTERM, signal.SIGINT):
        loop.add_signal_handler(sig, on_signal)

    async with server:
        await stop.wait()

    cell.heartbeat.stop()
    print("oncell runtime stopped", flush=True)


def main():
    parser = argparse.ArgumentParser(description="oncell runtime")
    parser.add_argument("--agent", required=True, help="path to agent.py")
    parser.add_argument("--cell-id", required=True, help="cell ID")
    parser.add_argument("--port", type=int, default=8080, help="listen port")
    parser.add_argument("--cells-dir", default="/cells", help="base cells directory")
    parser.add_argument("--control-plane-url", default=None, help="control plane URL for heartbeat")
    args = parser.parse_args()

    # Override from env
    cell_id = os.environ.get("ONCELL_CELL_ID", args.cell_id)
    port = int(os.environ.get("ONCELL_PORT", args.port))
    cells_dir = os.environ.get("ONCELL_CELLS_DIR", args.cells_dir)
    control_plane_url = os.environ.get("ONCELL_CONTROL_PLANE_URL", args.control_plane_url)

    agent = load_agent(args.agent)
    cell = Cell(cell_id, base_dir=cells_dir, control_plane_url=control_plane_url)

    asyncio.run(run_server(agent, cell, port))


if __name__ == "__main__":
    main()

File: typescript/src/runtime.ts
/**
 * Runtime — runs inside the cell. Loads the agent, handles requests.
 *
 * Usage (called by the base image entrypoint):
 *     npx tsx oncell/runtime.ts --agent /app/agent.ts --cell-id acme-corp --port 8080
 */

import { createServer, IncomingMessage, ServerResponse } from "http";
import { resolve } from "path";
import { Cell } from "./cell.js";
import { Agent } from "./agent.js";

async function loadAgent(agentPath: string): Promise<Agent> {
  const absPath = resolve(agentPath);
  const mod = await import(absPath);
  const exported = mod.default ?? Object.values(mod).find(
    (v: any) => typeof v === "function" && v.prototype instanceof Agent
  );
  if (!exported) {
    throw new Error(`no Agent subclass found in ${agentPath}`);
  }
  return new exported();
}

function readBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    req.on("data", (c) => chunks.push(c));
    req.on("end", () => resolve(Buffer.concat(chunks).toString()));
    req.on("error", reject);
  });
}

async function handleRequest(
  agent: Agent,
  cell: Cell,
  req: IncomingMessage,
  res: ServerResponse
) {
  const url = req.url ?? "/";
  let status = 200;
  let result: any;

  try {
    if (url === "/health") {
      result = { status: "ok", cell_id: cell.id };

    } else if (url === "/setup" && req.method === "POST") {
      await agent.setup(cell);
      result = { status: "ok" };

    } else if (url === "/request" && req.method === "POST") {
      const raw = await readBody(req);
      const body = JSON.parse(raw);
      if (!body.method) {
        status = 400;
        result = { error: "missing 'method' in request body" };
      } else {
        // Built-in DB entity CRUD — handled by the runtime, not the agent
        const dbResult = await handleDbMethod(cell, body.method, body.params ?? {});
        if (dbResult !== undefined) {
          result = dbResult;
        } else {
          result = await agent.onRequest(cell, body.method, body.params ?? {});
        }
      }

    } else if (url === "/teardown" && req.method === "POST") {
      await agent.teardown(cell);
      result = { status: "ok" };

    } else {
      status = 404;
      result = { error: `not found: ${url}` };
    }
  } catch (err: any) {
    status = 500;
    result = { error: err.message, stack: err.stack };
  }

  const body = JSON.stringify(result);
  res.writeHead(status, { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(body) });
  res.end(body);
}

/**
 * Built-in methods handled by the runtime before dispatching to the agent.
 *
 * DB entity CRUD (db_create, db_query, etc.) proxies to the cell's own
 * /api/db HTTP endpoint so that SDK calls and window.db calls in the
 * generated app share the same .data/db.json storage file.
 *
 * Key-value (db_set, db_get) uses cell.db directly (separate .data/cell.json).
 * File operations use cell.store directly.
 *
 * Returns undefined if the method is not built-in (falls through to agent).
 */

// Port of the cell's Next.js app — used for proxying entity CRUD
const CELL_APP_PORT = process.env.PORT || process.env.CELL_APP_PORT || "3000";
const CELL_APP_URL = `http://localhost:${CELL_APP_PORT}`;

async function cellDbFetch(method: string, path: string, body?: any): Promise<any> {
  const res = await fetch(`${CELL_APP_URL}/api/db${path}`, {
    method,
    headers: body ? { "Content-Type": "application/json" } : undefined,
    body: body ? JSON.stringify(body) : undefined,
  });
  return res.json();
}

async function handleDbMethod(cell: Cell, method: string, params: Record<string, any>): Promise<any | undefined> {
  switch (method) {
    // ─── Key-value (uses cell.db — separate storage) ───
    case "db_set":
      await cell.db.set(params.key, params.value);
      return { ok: true };
    case "db_get":
      return { value: await cell.db.get(params.key) };
    case "db_delete":
      await cell.db.delete(params.key);
      return { ok: true };
    case "db_keys":
      return { keys: await cell.db.keys(params.prefix || "") };

    // ─── Entity CRUD (proxies to cell's /api/db → shared db.json) ───
    case "db_create":
      return cellDbFetch("POST", "", { entity: params.entity, data: params.data });

    case "db_query": {
      const qs = new URLSearchParams({ entity: params.entity });
      if (params.where) qs.set("filter", JSON.stringify(params.where));
      if (params.orderBy) qs.set("sort", params.orderBy);
      if (params.order) qs.set("order", params.order);
      if (params.limit) qs.set("limit", String(params.limit));
      if (params.offset) qs.set("offset", String(params.offset));
      return cellDbFetch("GET", `?${qs.toString()}`);
    }

    case "db_get_all": {
      return cellDbFetch("GET", `?entity=${encodeURIComponent(params.entity)}`);
    }

    case "db_get_by_id":
      return cellDbFetch("GET", `?entity=${encodeURIComponent(params.entity)}&id=${encodeURIComponent(params.id)}`);

    case "db_update":
      return cellDbFetch("PUT", "", { entity: params.entity, id: params.id, data: params.data });

    case "db_delete_record":
      return cellDbFetch("DELETE", `?entity=${encodeURIComponent(params.entity)}&id=${encodeURIComponent(params.id)}`);

    // ─── File operations ───
    case "write_file":
      cell.store.write(params.path, params.content);
      return { ok: true };
    case "read_file":
      return { content: cell.store.read(params.path) };
    case "list_files":
      return { files: cell.store.list(params.path) };

    default:
      return undefined; // Not a built-in method — fall through to agent
  }
}

async function main() {
  const args = process.argv.slice(2);
  const getArg = (name: string, fallback: string) => {
    const idx = args.indexOf(name);
    return idx >= 0 ? args[idx + 1] : fallback;
  };

  const agentPath = process.env.ONCELL_AGENT_PATH ?? getArg("--agent", "/app/agent.ts");
  const cellId = process.env.ONCELL_CELL_ID ?? getArg("--cell-id", "default");
  const port = parseInt(process.env.ONCELL_PORT ?? getArg("--port", "8080"));
  const cellsDir = process.env.ONCELL_CELLS_DIR ?? getArg("--cells-dir", "/cells");
  const controlPlaneUrl = process.env.ONCELL_CONTROL_PLANE_URL ?? getArg("--control-plane-url", "");

  const agent = await loadAgent(agentPath);
  const cell = new Cell(cellId, { baseDir: cellsDir, controlPlaneUrl: controlPlaneUrl || undefined });

  const server = createServer((req, res) => {
    handleRequest(agent, cell, req, res).catch((err) => {
      res.writeHead(500, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ error: err.message }));
    });
  });

  cell.heartbeat.start();

  server.listen(port, "0.0.0.0", () => {
    console.log(`oncell runtime started — cell=${cellId} port=${port}`);
  });

  const shutdown = () => {
    console.log("shutting down...");
    cell.heartbeat.stop();
    server.close(() => {
      console.log("oncell runtime stopped");
      process.exit(0);
    });
  };

  process.on("SIGTERM", shutdown);
  process.on("SIGINT", shutdown);
}

main().catch((err) => {
  console.error("runtime failed:", err);
  process.exit(1);
});
